packages/blocks/image/src/components/Media.tsx

```diff
@@ -0,0 +1,366 @@
+import {
+  BlockProtocolEntity,
+  BlockProtocolLinkGroup,
+  BlockProtocolUpdateEntitiesAction,
+  BlockProtocolUploadFileFunction,
+} from "blockprotocol";
+import { BlockComponent } from "blockprotocol/react";
+import React, {
+  Dispatch,
+  SetStateAction,
+  useCallback,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from "react";
+import { unstable_batchedUpdates } from "react-dom";
+import { ErrorAlert } from "./ErrorAlert";
+import { MediaWithCaption } from "./MediaWithCaption";
+import { UploadMediaForm } from "./UploadMediaForm";
+
+// https://www.typescriptlang.org/docs/handbook/release-notes/overview.html#recursive-conditional-types
+type Awaited<T> = T extends PromiseLike<infer U> ? Awaited<U> : T;
+type FileType = Awaited<ReturnType<BlockProtocolUploadFileFunction>>;
+type UpdateEntitiesMediaAction = {
+  initialCaption?: string;
+  initialWidth?: number;
+  file?: FileType;
+};
+
+const useDefaultState = <T extends any>(
+  defaultValue: T,
+): [T, Dispatch<SetStateAction<T>>] => {
+  const defaultStateValue = {
+    prevDefault: defaultValue,
+    currentValue: defaultValue,
+  };
+  const [{ prevDefault, currentValue }, setNextValue] =
+    useState(defaultStateValue);
+
+  if (prevDefault !== defaultValue) {
+    setNextValue(defaultStateValue);
+  }
+
+  const setState = useCallback((value: SetStateAction<T>) => {
+    setNextValue((prevValue) => {
+      const nextValue =
+        // @ts-expect-error We know this is callable, but TS thinks value
+        // could be another kind of function
+        typeof value === "function" ? value(prevValue.currentValue) : value;
+
+      return {
+        ...prevValue,
+        currentValue: nextValue,
+      };
+    });
+  }, []);
+
+  return [currentValue, setState];
+};
+
+function getLinkGroup(params: {
+  linkGroups: BlockProtocolLinkGroup[];
+  path: string;
+  sourceEntityId: string;
+}): BlockProtocolLinkGroup | undefined {
+  const { linkGroups, path, sourceEntityId } = params;
+
+  const matchingLinkGroup = linkGroups.find(
+    (linkGroup) =>
+      linkGroup.path === path && linkGroup.sourceEntityId === sourceEntityId,
+  );
+
+  return matchingLinkGroup;
+}
+
+function getLinkedEntities<T>(params: {
+  sourceEntityId: string;
+  path: string;
+  linkGroups: BlockProtocolLinkGroup[];
+  linkedEntities: BlockProtocolEntity[];
+}): (BlockProtocolEntity & T)[] | null {
+  const { sourceEntityId, path, linkGroups, linkedEntities } = params;
+
+  const matchingLinkGroup = getLinkGroup({
+    linkGroups,
+    path,
+    sourceEntityId,
+  });
+
+  if (!matchingLinkGroup) {
+    return null;
+  }
+
+  const destinationEntityId = matchingLinkGroup.links[0].destinationEntityId;
```

Comment:
Appreciate you only moved this, but might be worth handling checking this index access now so we don't have to when we introduce the tsconfig flag

Reply:
👍